pymatviz/colormaps.py

```diff
@@ -0,0 +1,122 @@
+"""Generate custom colormaps."""
+
+
+from __future__ import annotations
+
+import matplotlib
+import numpy as np
+from matplotlib.colors import Colormap, LinearSegmentedColormap, ListedColormap
+
+
+def combine_two(
+    cmaps: list[Colormap | str],
+    node: float = 0.5,
+    N: int = 256,
+    reverse: bool = False,
+) -> Colormap:
+    """Create a composite matplotlib Colormap by combining two colormaps.
+
+    This function takes a list of two colormaps (or their names as strings)
+    and creates a composite colormap by blending them together.
+
+    Parameters:
+    - cmaps (List[Colormap | str]): A list containing two colormaps or
+        their names as strings.
+    - node (float, optional): The blending point between the two colormaps,
+        in range [0, 1]. Defaults to 0.5.
+    - N (int, optional): The number of RGB quantization levels.
+        Defaults to 256.
+    - reverse (bool, optional): Reverse the resulting colormap.
+
+    Returns:
+    - Colormap: The composite colormap.
+
+    Raises:
+    - TypeError: If the list contains mixed datatypes or
+        invalid colormap names.
+
+    Note:
+    - The colormaps are combined from bottom to top by default.
+
+    References:
+    - https://matplotlib.org/stable/users/explain/colors/colormap-manipulation.html
+    - https://stackoverflow.com/questions/31051488/combining-two-matplotlib-colormaps/31052741#31052741
+
+    Examples:
+    >>> combined_cmap = combine_two_colormaps(['viridis', 'plasma'],
+            node=0.3, N=128, reverse=True)
+
+    Todo:
+    - Generalize to a list of Colormaps.
+    """
+    # Check cmap datatype and convert to List[Colormap]
+    if all(isinstance(cmap, str) for cmap in cmaps):
+        cmaps = [matplotlib.colormaps[s] for s in cmaps]
+    elif any(not isinstance(cmap, Colormap) for cmap in cmaps):
+        raise TypeError("Invalid datatype. Expect either all Colormaps or all strings.")
+
+    # Sample two source colormaps
+    cmap_0 = cmaps[0](np.linspace(0, 1, int(N * node)))
+    cmap_1 = cmaps[1](np.linspace(0, 1, N - int(N * node)))
+
+    # Merge Colormaps
+    if reverse:
+        return LinearSegmentedColormap.from_list(
+            "composite_cmap_r", np.vstack((cmap_0, cmap_1))
+        ).reversed()
+    else:
+        return LinearSegmentedColormap.from_list(
+            "composite_cmap", np.vstack((cmap_0, cmap_1))
+        )
+
+
+def truncate(
+    cmap: str | Colormap,
+    start: float,
+    end: float,
+    N: int = 256,
+) -> Colormap:
+    """Truncate a matplotlib Colormap to a specified range.
```

Comment:
same here, i think we could use `colors.LinearSegmentedColormap.from_list` instead (see https://stackoverflow.com/a/18926541)

Reply:
take your time to recover! no rush

Reply:
Hi @janosh I'm back and feel good. This `cmasher` package looks good and is able to handle custom colormaps, so I would drop the `truncate` function, and also see if I could find an already available package for the `combine_two` function as well.

```python
import cmasher as cmr
import matplotlib.pyplot as plt

# Create another custom cmap from custom cmap
built_in_cmap = plt.get_cmap('viridis')
custom_cmap = cmr.get_sub_cmap('viridis', 0.1, 0.9)

new_custom_cmap = cmr.get_sub_cmap(custom_cmap, 0.1, 0.3)

# Generate some data to plot
x = np.random.rand(100)
y = np.random.rand(100)
z = x**2+y**2

# Make scatter plot of data with colormap
plt.scatter(x, y, c=z, cmap=new_custom_cmap, s=300)
plt.show()
```